### PR TITLE
DAOS-5734 dfs: fix path length check

### PR DESCRIPTION
**Pull request by mchaarawi**
as part of SDL, we switched to use strnlen instead of strlen, but
overlooked the fact that strnlen does not return more than max_len.
So the check for len should be == not > otherwise the error case would
not be triggered.

Add other NULL input argument checks.

Signed-off-by: Mohamad Chaarawi <mohamad.chaarawi@intel.com>

### CODE REVIEW

**Comment by daltonbohning** [comment]
This fixes the issue, but seeing it a second time, I think I overlooked another off-by-error when I originally wrote up the ticket.
We are checking `strnlen(prefix, PATH_MAX-1) == PATH_MAX-1`. This means that the prefix can _**contain**_ at most `PATH_MAX-2` characters, because `PATH_MAX-1` is too much. But we only need one character for the terminator '\0', so really I think we could allow `PATH_MAX-1` characters, but error on `PATH_MAX` characters.

I'm noticing in master `dfs.c`, we have 3 of these such checks in place:

_**`open_symlink`**_ - This one is incorrect
https://github.com/daos-stack/daos/blob/fa6d7e03b1c82a2de431d9640b712efb56165edd/src/client/dfs/dfs.c#L827

_**`dfs_set_prefix`**_ - This one looks to be the most correct
https://github.com/daos-stack/daos/blob/fa6d7e03b1c82a2de431d9640b712efb56165edd/src/client/dfs/dfs.c#L1502

_**`dfs_lookup`**_ - This is the one you just fixed. Previously similar to `open_symlink`
https://github.com/daos-stack/daos/blob/fa6d7e03b1c82a2de431d9640b712efb56165edd/src/client/dfs/dfs.c#L1797

I think the one in `dfs_set_prefix` is the most correct:
`strnlen(prefix, PATH_MAX) > PATH_MAX-1)`
This one _**looks for**_ up to `PATH_MAX`, so it can know if we overflowed `PATH_MAX-1`, leaving one character for '\0'.
Whereas if we _**look for**_ `PATH_MAX-1`, then we only know if we overflowed `PATH_MAX-2`, leaving two characters for '\0'.

Sorry for not noticing this before. What do you think?

File: src/client/dfs/dfs.c [hunk 1]
```diff
@@ -571,7 +571,7 @@ check_name(const char *name)
 {
 	if (name == NULL || strchr(name, '/'))
 		return EINVAL;
-	if (strnlen(name, DFS_MAX_PATH) > DFS_MAX_PATH)
+	if (strnlen(name, DFS_MAX_PATH) == DFS_MAX_PATH)
```
Comment (daltonbohning) [inline src/client/dfs/dfs.c:574]:
```suggestion
	if (strnlen(name, DFS_MAX_PATH+1) > DFS_MAX_PATH)
```

File: src/client/dfs/dfs.c [hunk 2]
```diff
@@ -824,7 +824,8 @@ open_symlink(dfs_t *dfs, daos_handle_t th, dfs_obj_t *parent, int flags,
 	int			rc;
 
 	if (flags & O_CREAT) {
-		if (value == NULL || strnlen(value, PATH_MAX-1) > PATH_MAX-1)
+		if (value == NULL ||
+		    strnlen(value, PATH_MAX - 1) == PATH_MAX - 1)
```
Comment (daltonbohning) [inline src/client/dfs/dfs.c:828]:
```suggestion
		    strnlen(value, PATH_MAX) > PATH_MAX - 1)
```

File: src/client/dfs/dfs.c [hunk 3]
```diff
@@ -1499,7 +1506,7 @@ dfs_set_prefix(dfs_t *dfs, const char *prefix)
 		return 0;
 	}
 
-	if (prefix[0] != '/' || strnlen(prefix, PATH_MAX) > PATH_MAX-1)
+	if (prefix[0] != '/' || strnlen(prefix, PATH_MAX - 1) == PATH_MAX - 1)
```
Comment (daltonbohning) [inline src/client/dfs/dfs.c:1509]:
```suggestion
	if (prefix[0] != '/' || strnlen(prefix, PATH_MAX) > PATH_MAX - 1)
```

File: src/client/dfs/dfs.c [hunk 4]
```diff
@@ -1794,7 +1801,7 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
 		return EINVAL;
 	if (_obj == NULL)
 		return EINVAL;
-	if (path == NULL || strnlen(path, PATH_MAX-1) > PATH_MAX-1)
+	if (path == NULL || strnlen(path, PATH_MAX - 1) == PATH_MAX - 1)
```
Comment (daltonbohning) [inline src/client/dfs/dfs.c:1804]:
```suggestion
	if (path == NULL || strnlen(path, PATH_MAX) > PATH_MAX - 1)
```

**Comment by mchaarawi** [comment]
> This fixes the issue, but seeing it a second time, I think I overlooked another off-by-error when I originally wrote up the ticket.
> We are checking `strnlen(prefix, PATH_MAX-1) == PATH_MAX-1`. This means that the prefix can _**contain**_ at most `PATH_MAX-2` characters, because `PATH_MAX-1` is too much. But we only need one character for the terminator '\0', so really I think we could allow `PATH_MAX-1` characters, but error on `PATH_MAX` characters.
> 
> I'm noticing in master `dfs.c`, we have 3 of these such checks in place:
> 
> _**`open_symlink`**_ - This one is incorrect
> https://github.com/daos-stack/daos/blob/fa6d7e03b1c82a2de431d9640b712efb56165edd/src/client/dfs/dfs.c#L827
> 
> _**`dfs_set_prefix`**_ - This one looks to be the most correct
> https://github.com/daos-stack/daos/blob/fa6d7e03b1c82a2de431d9640b712efb56165edd/src/client/dfs/dfs.c#L1502
> 
> _**`dfs_lookup`**_ - This is the one you just fixed. Previously similar to `open_symlink`
> https://github.com/daos-stack/daos/blob/fa6d7e03b1c82a2de431d9640b712efb56165edd/src/client/dfs/dfs.c#L1797
> 
> I think the one in `dfs_set_prefix` is the most correct:
> `strnlen(prefix, PATH_MAX) > PATH_MAX-1)`
> This one _**looks for**_ up to `PATH_MAX`, so it can know if we overflowed `PATH_MAX-1`, leaving one character for '\0'.
> Whereas if we _**look for**_ `PATH_MAX-1`, then we only know if we overflowed `PATH_MAX-2`, leaving two characters for '\0'.
> 
> Sorry for not noticing this before. What do you think?

strnlen() does not include the NULL terminator, i don't think.

File: src/client/dfs/dfs.c [hunk 5]
```diff
@@ -824,7 +824,8 @@ open_symlink(dfs_t *dfs, daos_handle_t th, dfs_obj_t *parent, int flags,
 	int			rc;
 
 	if (flags & O_CREAT) {
-		if (value == NULL || strnlen(value, PATH_MAX-1) > PATH_MAX-1)
+		if (value == NULL ||
+		    strnlen(value, PATH_MAX - 1) == PATH_MAX - 1)
 			return -DER_INVAL;
```
Comment (berserk-fury) [inline src/client/dfs/dfs.c:829]: This is not part of this patch. I think that we are mixing errno codes and DER_ codes on this function

`return daos_der2errno(-DER_INVAL)`
or
`return EINVAL;`

Comment (mchaarawi) [reply src/client/dfs/dfs.c:829]: yes you are right.

**Comment by daltonbohning** [comment]
> strnlen() does not include the NULL terminator, i don't think.

Correct, strnlen does not include the NULL terminator. 
Let's say `PATH_MAX=4`.
Then the largest string we could allow would have 3 characters + '\0', for a total of 4 characters.
And this string would be valid: `prefix = "123\0"`
And `strnlen(prefix, PATH_MAX-1)` would return `3`. And `3==PATH_MAX-1` is `true`, so in this case we would error on a string that is in fact valid.
If we allow `PATH_MAX` characters, but reserve one character for `'\0'`, then I think we need one of these checks:
`strnlen(prefix, PATH_MAX) > PATH_MAX-1`
`strnlen(prefix, PATH_MAX) == PATH_MAX`

**Comment by berserk-fury** [comment]
Do we have some unit test to verify that strnlen() extra protection works as expected?
As Dalton pointed out, there are some corner cases that could break in the future.

**Comment by mchaarawi** [comment]
> > This fixes the issue, but seeing it a second time, I think I overlooked another off-by-error when I originally wrote up the ticket.
> > We are checking `strnlen(prefix, PATH_MAX-1) == PATH_MAX-1`. This means that the prefix can _**contain**_ at most `PATH_MAX-2` characters, because `PATH_MAX-1` is too much. But we only need one character for the terminator '\0', so really I think we could allow `PATH_MAX-1` characters, but error on `PATH_MAX` characters.
> > I'm noticing in master `dfs.c`, we have 3 of these such checks in place:
> > _**`open_symlink`**_ - This one is incorrect
> > https://github.com/daos-stack/daos/blob/fa6d7e03b1c82a2de431d9640b712efb56165edd/src/client/dfs/dfs.c#L827
> > 
> > _**`dfs_set_prefix`**_ - This one looks to be the most correct
> > https://github.com/daos-stack/daos/blob/fa6d7e03b1c82a2de431d9640b712efb56165edd/src/client/dfs/dfs.c#L1502
> > 
> > _**`dfs_lookup`**_ - This is the one you just fixed. Previously similar to `open_symlink`
> > https://github.com/daos-stack/daos/blob/fa6d7e03b1c82a2de431d9640b712efb56165edd/src/client/dfs/dfs.c#L1797
> > 
> > I think the one in `dfs_set_prefix` is the most correct:
> > `strnlen(prefix, PATH_MAX) > PATH_MAX-1)`
> > This one _**looks for**_ up to `PATH_MAX`, so it can know if we overflowed `PATH_MAX-1`, leaving one character for '\0'.
> > Whereas if we _**look for**_ `PATH_MAX-1`, then we only know if we overflowed `PATH_MAX-2`, leaving two characters for '\0'.
> > Sorry for not noticing this before. What do you think?
> 
> strnlen() does not include the NULL terminator, i don't think.

actually you are correct because the string at max size does exactly than and ignores the NULL terminator and errors out.

**Comment by mchaarawi** [comment]
> > strnlen() does not include the NULL terminator, i don't think.
> 
> Correct, strnlen does not include the NULL terminator.
> Let's say `PATH_MAX=4`.
> Then the largest string we could allow would have 3 characters + '\0', for a total of 4 characters.
> And this string would be valid: `prefix = "123\0"`
> And `strnlen(prefix, PATH_MAX-1)` would return `3`. And `3==PATH_MAX-1` is `true`, so in this case we would error on a string that is in fact valid.
> If we allow `PATH_MAX` characters, but reserve one character for `'\0'`, then I think we need one of these checks:
> `strnlen(prefix, PATH_MAX) > PATH_MAX-1`
> `strnlen(prefix, PATH_MAX) == PATH_MAX`

yes right, thanks
we need to add some tests :-)

**Comment by mchaarawi** [comment]
> Do we have some unit test to verify that strnlen() extra protection works as expected?
> As Dalton pointed out, there are some corner cases that could break in the future.

no we should add a lot more unit tests to DFS. we don't have someone to work on this currently.
hopefully soon

**Comment by daltonbohning** [comment]
> Do we have some unit test to verify that strnlen() extra protection works as expected?
> As Dalton pointed out, there are some corner cases that could break in the future.

We do have this function:
https://github.com/daos-stack/daos/blob/fa6d7e03b1c82a2de431d9640b712efb56165edd/src/client/dfs/dfs.c#L569-L577

Maybe we can use this (or something similar), and wrap each of these checks with that function, and then a unit test for that function?

**Comment by daltonbohning** [comment]
> no we should add a lot more unit tests to DFS. we don't have someone to work on this currently.
> hopefully soon

Ah, I see.

**Review by daltonbohning — APPROVED** [review]
I think this is good :)
Thanks!

**Review by berserk-fury — APPROVED** [review]
LGTM

**Comment by mchaarawi** [comment]
Failed tests are from DAOS-5763
DFS is not called from any of those failing tests.

**Comment by jolivier23** [comment]
What is the fio failure?
